### ISSUE TITLE
wrong_self_convention triggers in trait method implementations

### ISSUE THREAD

**Issue by kangalioo** [rust-lang/rust-clippy#7123]
Lint name: wrong_self_convention

I tried this code:

```rust
struct Error {}

trait FromMeta: Sized {
    fn from_char(_value: char) -> Result<Self, Error> { Err(Error {}) }
    fn from_bool(_value: bool) -> Result<Self, Error> { Err(Error {}) }
}

impl FromMeta for char {
    fn from_char(value: Self) -> Result<Self, Error> { Ok(value) }
}

impl FromMeta for bool {
    fn from_bool(value: Self) -> Result<Self, Error> { Ok(value) }
}
```

I expected to see this happen: nothing is linted

Instead, this happened: the from_char implementation for char and from_bool implementation for bool were linted because Clippy thinks the `value: Self` parameter is a `self` parameter and `from_*` methods shouldn't take self parameters.

However, `FromMeta` is a trait and will be implemented by lots of different types. On some types, the trait names may seem slightly awkward, but I think there is not much the programmer can realistically do about that. I think it would be beneficial if this lint weren't triggered in trait implementations.

Context: the [`FromMeta`](https://github.com/TedDriggs/darling/blob/cd0dbbacd961c3ef41acce24875f0de9e5ba3140/core/src/from_meta.rs#L49-L127) trait is similar in spirit to serde's Deserialize. It comes from the darling crate, a proc-macro attribute parser. Links for the offending trait implementations: [bool](https://github.com/TedDriggs/darling/blob/cd0dbbacd961c3ef41acce24875f0de9e5ba3140/core/src/from_meta.rs#L142-L144) and [char](https://github.com/TedDriggs/darling/blob/cd0dbbacd961c3ef41acce24875f0de9e5ba3140/core/src/from_meta.rs#L160-L162).

### Meta

- `cargo clippy -V`: clippy 0.1.53 (b849326 2021-04-21)
- `rustc -Vv`:
```
rustc 1.53.0-nightly (b84932674 2021-04-21)
binary: rustc
commit-hash: b849326744a8eec939e592f0ab13bff85cc865d3
commit-date: 2021-04-21
host: x86_64-unknown-linux-gnu
release: 1.53.0-nightly
LLVM version: 12.0.0
```

**Comment by kangalioo** [comment]
This seems to be a duplicate of #7032. Sorry for not searching for duplicates before posting
